csrc/device_lower/analysis/predicate_elimination.cpp

```diff
@@ -249,9 +252,9 @@ class PredicateChcker : public IterVisitor {
     //  logic. But this part along with [Predicate Inversion for CpAsync]
     //  should be cleaned up all together when we extend predicate/masking
     //  logic to cover this usage.
-    NVF_ERROR(
-        !(ir_utils::isCpAsyncOp(expr) && needs_predicate_smem_access),
-        "predicate removal: unsupported use case of cp.async");
+    // NVF_ERROR(
```

Comment:
Hi @zasdfgbnm,  in innerOuter persistent scheduler, I am trying to use `CpAsync` to avoid using tmp registers when copying data from gmem to smem. But I need to remove this restriction. Do you know why we have this restriction? The comment seems refering to `NVFuserTest.FusionCpAsyncPredicate_CUDA` but it still passes the test without this restriction. Thanks!

Reply:
I added a test and the generated PTX is
```
        asm volatile(
          "{\n"
          "  .reg .pred p0; \n"
          "  setp.ne.b32 p0, %3, 0;\n"
          "  cp.async.ca.shared.global [%0], [%1], %2, p0;\n"
          "}\n"
          :
          :"r"((uint32_t)((i11 + (i12 * i36)))),
           "l"((ptr33 + i37)),
           "n"(16LL),
           "n"((uint32_t)(false))
        );
```
p0 is set to `false`, but the data is still correctly loaded from global to shared as the test passed correctly. Is this expected?

Reply:
I believe the problem is like this, first of all, `cp.async` has built in predicate, so let's consider `cp_async` as the following function:
```C++
template<typename T>
void cp_async(bool ignore_src, T* dest, T* src) {
  if (ignore_src) {
    *dest = 0;
  } else {
    *dest = *src;
  }
}
```

Now, let's consider the following fusion:

```
gmem_tv: shape = [1, 8]
smem_tv = cp_async(gmem_tv);
...
```
and let's schedule the fusion like this:

```C++
smem_tv->split(0, 2);
// smem_tv[1, 2, 8];
smem_tv->axis(1)->parallelize(TIDx);
// smem_tv[1, TIDx, 8];
```

for this schedule, the index of `gmem_tv` and `smem_tv` will both be as simple as `threadIdx.x * 8`.

Now, suppose that, we launch kernel with `blockDim.x = 2`, then this schedule will be lowered as:

```C++
cp_async(!(threadIdx.x * 8 < 8), smem_tv + threadIdx.x * 8, gmem_tv + threadIdx.x * 8);
```

this is fine, it will work correctly.

However, let's consider that now, instead of launching kernel with `blockDim.x = 2`, we launch with `blockDim.x = 256`. For this case, in the fusion, there will be some IterDomain with extent `256` parallelized on `TIDx`. For this case, we will lower the kernel as:

```C++
cp_async(!(threadIdx.x * 8 < 8 && threadIdx.x < 2) , smem_tv + threadIdx.x * 8, gmem_tv + threadIdx.x * 8);
```

But is this correct? No!!! `cp_async` will write to `smem_tv + threadIdx.x * 8` regardless of the value of the predicate. The predicate just changes the value that `cp_async` writes, but not whether to write or now. Due to this nature of `cp_async`, we will have an out of boundary access. The correct code to generate should be:

```C++
if (threadIdx.x < 2) {
  cp_async(!(threadIdx.x * 8 < 8) , smem_tv + threadIdx.x * 8, gmem_tv + threadIdx.x * 8);
}
```

but we are not generating code this way.

Reply:
> p0 is set to `false`, but the data is still correctly loaded from global to shared as the test passed correctly. Is this expected?

The predicate of `cp.async` is `ignore-src`, so you need `true` predicate to not loading from gmem.

Reply:
Thanks for the detailed demo. It seems that the test case I added is generating the correct code. Without using cp.async:
```
     if ((b28 && (i20 < (-i30)))) {
        loadGeneric<float, 4>( &T5[(i6 + (i9 * i29))],  &T0[(i27 + i30)]);
      }
```
with cp.async:
```
      if ((b35 && (i25 < (-i37)))) {
        asm volatile(
          "{\n"
          "  .reg .pred p0; \n"
          "  setp.ne.b32 p0, %3, 0;\n"
          "  cp.async.ca.shared.global [%0], [%1], %2, p0;\n"
          "}\n"
          :
          :"r"((uint32_t)((i11 + (i12 * i36)))),
           "l"((ptr33 + i37)),
           "n"(16LL),
           "n"((uint32_t)(false))
        );
      }
```
Let me check how to trigger the error you listed.

Reply:
@zasdfgbnm I can reproduce the error you mensioned with a simple case where TIDx is set to 48 but only needs 12 to load from gmem to smem.
```
TEST_F(NVFuserTest, FusionCpAsyncPredicateError_CUDA) {
  Fusion fusion;
  FusionGuard fg(&fusion);

  // Using vectorization so need to keep n multiple of 4.
  int m = 33, n = 48;

  TensorView* tv0 = makeContigTensor(2);
  fusion.addInput(tv0);
  auto tv1 = exp(tv0);
  fusion.addOutput(tv1);

  auto lst = LoadStoreOpType::Set;
  if(std::getenv("USE_CPASYNC")){
    lst = LoadStoreOpType::CpAsync;
  }
  auto tvs = tv0->cacheAfter(lst);
  tvs->setMemoryType(MemoryType::Shared);

  tvs->split(-1, 4);
  tvs->axis(-1)->parallelize(ParallelType::Vectorize);
  tvs->axis(-2)->parallelize(ParallelType::TIDx);
  tvs->axis(-3)->parallelize(ParallelType::BIDx);

  tv1->axis(-1)->parallelize(ParallelType::TIDx);
  tv1->axis(-2)->parallelize(ParallelType::BIDx);

  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
  at::Tensor t0 = at::randn({m, n}, options);

  FusionExecutor fe;
  if (!deviceMajorMinorCheck(8)) {
    ASSERT_THAT(
        [&]() { fe.compileFusion(&fusion, {t0}); },
        testing::ThrowsMessage<nvfuser::nvfError>(testing::HasSubstr(
            "Reason: LoadStoreOpType::CpAsync requires Ampere")));
    GTEST_SKIP() << "skipping tests on pre-AMPERE GPUs";
  } else {
    fe.compileFusion(&fusion, {t0});
  }

  fe.compileFusion(&fusion, {t0});
  auto cg_outputs = fe.runFusion({t0});
  auto ref = t0.exp();
  testValidate(&fusion, cg_outputs, {t0}, {ref}, __LINE__, __FILE__);
}
```

Reply:
> I can reproduce the error you mensioned with a simple case where TIDx is set to 48 but only needs 12 to load from gmem to smem.

I don't have a good suggestion on how to proceed with this PR, but could you open a separate PR committing your test, and open an issue?